### ISSUE TITLE
Meson doesn't detect files ending in .C as C++

### ISSUE THREAD

**Issue by jml1795** [mesonbuild/meson#4919]
It seems that meson doesn't correctly detect files ending in the extension `.C` (note capital C) as C++ code. According to the [Gcc manual](https://gcc.gnu.org/onlinedocs/gcc-4.8.5/gcc/Overall-Options.html#index-file-name-suffix-71), this is an accepted file extension to be processed as C++. I realize that it isn't a common extension nor is it one I have ever seen outside my current project, but I cannot change the convention in my project at this time. Any chance we could add support for this in meson? I am happy to make the update if you can point me in the right direction.

I can get around this by passing `-xc++` but it would be nice if we could support it directly.

This can be demonstrated by running `meson init -l cpp` and changing the generated file's extension to `.C`.

I am running gcc 4.8.5 if that matters and meson off `07f1175c`. Thanks again.

**Comment by QuLogic** [comment]
Generally the problem with this is that other systems are not necessarily case-sensitive. There are two incomplete PRs for this that never hit those corner cases #3409, #3415.

**Comment by jml1795** [comment]
@QuLogic Thanks for the links. Is there a simple solution we can implement? Something to the effect of @nirbheek 's [proposal](https://github.com/mesonbuild/meson/pull/3409#discussion_r181566229) to be case sensitive on extensions and platform combinations where it matters and follow existing behavior elsewhere?

**Comment by jpakkane** [comment]
Because of backwards compatibility, everything is terrible. There are Windows projects that have `.C` ending for C files because they originate from those systems. Then there are other projects that have `.C` for C++. Then, thanks to the magic of subprojects, you may have _both_ of these within one project and it is impossible to know whether a file with the extension `.C` should be treated as C or C++.

**Comment by jml1795** [comment]
@jpakkane  Maybe overly simplistic, but could we allow a project-wide opt in? Something to the effect of:

```py
map_ext_to_language(ext: '.c', language: 'c')
map_ext_to_language(ext: '.C', language: 'cpp')
...
```
which would maintain current behavior if not overridden and override the extension detection mechanism if set. Not sure this solves the case of mixing conventions in subprojects, but maybe gets us closer?

**Comment by nirbheek** [comment]
It's really much simpler to just rename your source files.

**Comment by jpakkane** [comment]
It's also the "correct" long term solution and something you should do _anyway_. Granted, the transition does take some effort but in the end it's worth it.

**Comment by scivision** [comment]
As an external reference, [CMake handles this](https://cmake.org/cmake/help/latest/command/set_source_files_properties.html) via
```cmake
set_source_files_properties(file1.C file2.C PROPERTIES 
  LANGUAGE CXX)
```

**Comment by jml1795** [comment]
@nirbheek I realize it is a much better solution to rename, but this is not an option for this project. Unfortunately, we have about 7 million lines of C++ that adhere to this convention and doing a wholesale rename of every translation unit isn't going to get approved just to use a particular build system. I am happy to work on the change if we can come up with a design that meets all the corner cases.

I will test the `-xc++` flags some more, but I believe I encountered some linker issues that this couldn't resolve. I will report back.

**Comment by jpakkane** [comment]
You need to have at least one file that Meson recognizes as C++ (usually `main.cpp`) in a target so that Meson will link it as C++ and do all the necessary magic to make it work.

**Comment by nirbheek** [comment]
> Unfortunately, we have about 7 million lines of C++ that adhere to this convention

Wow, that's a lot of legacy code. Do tell us how your tests go!

**Comment by jml1795** [comment]
@nirbheek Yeah, not going to be fun

**Comment by mensinda** [comment]
Maybe a new kwarg for build targets to specify the language mapping on a per target basis?

Something like:
```meson
executable(..., language_mapping: {
  'c':   ['c'],
  'cpp': ['cpp', 'cxx', 'C']
})
```

**Comment by scivision** [comment]
this did not work with Meson master branch:
```meson

project('C++ exts', 'cpp')

executable('hello', 'main.cpp', 'abc.xyz')
```

where it would compile and run with `-xc++` flag using g++ or clang++

> ERROR: No specified compiler can handle file abc.xyz

**Comment by nirbheek** [comment]
One simple solution I can think of is to add a `language:` kwarg to the `files()` function.

**Comment by jpakkane** [comment]
Something like this should work:

```meson
project('C++ exts', 'cpp', 'c')

add_project_arguments('-lxc++', language: 'c')

executable('hello', 'main.cpp', 'something.C')
```

**Comment by scivision** [comment]
yes @nirbheek `files( ... , language : 'cpp')` would be even better than per-target

**Comment by scivision** [comment]
OK the trick is that the offending extension be something recognized. In this case luckily `.C` is recognized as C.  This meson.build works for me with GCC and Clang:
```meson
project('C++ exts', 'cpp', 'c')

executable('hello', 'main.cpp', 'abc.C')
```

and simply
```sh
meson ..
ninja
```

**Comment by xclaesse** [comment]
> One simple solution I can think of is to add a `language:` kwarg to the `files()` function.

+1, it can be generally useful IMHO.

**Comment by jml1795** [comment]
@jpakkane 
```py
add_project_arguments('-lxc++', language: 'c')
```
implies you cannot have c and c++ code in the same project, even if they aren't in the same target, correct?

@scivision 
```py
project('C++ exts', 'cpp', 'c')
executable('hello', 'main.cpp', 'abc.C')
```
works, provided you have at least one file w/ `.cpp` as an extension per target, correct?

**Comment by xclaesse** [comment]
In glib we had one objective-c file with `.c` extension instead of `.m` for historical reasons. Luckily in that case we were able to just rename it, but having the `language:` kwarg on `files()` would have been useful in the case we couldn't rename the file.

**Comment by scivision** [comment]
@jml1795 this works: a less-obtrusive workaround: create a new, empty file with .cpp extension and put it in every affected target--the file `blank.cpp` is literally a 0-byte empty file.
```meson
executable('main.C', 'abc.C', 'blank.cpp')
```

**Comment by scivision** [comment]
It would be preferable long-term for Meson to have a `files(..., language : 'cpp')` option for where the extension isn't recognized, for situations like @xclaesse notes

**Comment by nirbheek** [comment]
I just remembered that there is also the use-case where a mostly-C source file has `#ifdefs` with ObjC code in it that is only processed on macOS. In that case, it's cumbersome to conditionally change the extension.

**Comment by jpakkane** [comment]
> implies you cannot have c and c++ code in the same project, even if they aren't in the same target, correct?

Yes. If you need that, put the compile argument in the target's args rather than global.

**Comment by jml1795** [comment]
@jpakkane It looks like a per-target `-xc++` doesn't solve the linker problem, but does get the source to compile. I could manually add the particular system libraries I need to link against, but this is likely to be brittle particularly when going between compilers / platforms.

Short of not using the `.C` extension for C++ (a non-option in my prject at this time) or adding an empty TU in each target (which could result in literally thousands of additional compiler invocations) there doesn't seem to be a current workaround. It sounds like we could add a `language` option to the files list but there may be issues when embedding, e.g. ObjC inside C source - @nirbheek Is this correct? How is this case currently handled? I am sure there are other cases in addition to case-insensitive extension collisions where specifying the language in the file list could be useful. I'm assuming you had something like this in mind?

```py
cuda_src = files('a.cu')   # existing language detection
cpp_src = files('b.C': langauage: 'cpp')   # override language detection
lib = static_library('lib', cuda_src, cpp_src)
``` 

It seems we also detect whether we are on a case sensitive platform and then handle the language  decision based on that? For instance, if we are on a linux platform we select the language based on the raw extension, however, if we are on a case-insensitive platform, we select the language based on `ext.lower()`. This would maintain the existing behavior on platforms that are case-insensitive and add support for the full language mapping on those that support the difference. This would also allow meson to support projects w/ both c-source code under `.c` and c++-source under `.C`. 

I have the time to work this if we can hone in on a direction.

**Comment by jeandet** [comment]
> Because of backwards compatibility, everything is terrible. There are Windows projects that have `.C` ending for C files because they originate from those systems. Then there are other projects that have `.C` for C++. Then, thanks to the magic of subprojects, you may have _both_ of these within one project and it is impossible to know whether a file with the extension `.C` should be treated as C or C++.

@jpakkane, @nirbheek  I'm not sure this is a strong argument, the fact is that .C is C++ and the case insensitive issue apply to the whole name anyway. 
Today if you give 'A_FILENAME.c' to any meson function on a case insensitive system it will match 'a_filename.C' but will fail on any case sensitive system. 
So relying on case insensitive systems is wrong and I really don't think we should do any design choice which encourage people to ignore the case in file names. We should force them to make portable build defs (which works on case sensitive systems).

**Comment by xclaesse** [comment]
@jml1795 yes we should definitely support `files('foo.C', language : 'cpp')`. About objC use-case, it would work like that:
```meson
src = files(...)
if system == 'darwin'
  src += files('foo.c', language : 'objc')
else
  src += files('foo.c')
endif
executable('app', src)
```

**Comment by dcbaker** [comment]
Honestly, treating .c files as rust sources (or whatever other madness this is meant to support) is going to come back to bite us. I've worked on a project written in cmake that did something like (using meson psudo code because my cmake skills are weak)

```meson
f= files('a.m', 'b.m', 'c.m', language : 'c')
exectuable('foo', f, c_args : '-ObjC')
```

We shouldn't be enabling this kind of madness. If we're serious about cross platform we shouldn't support patterns that don't work on some platforms. The fact that on a fat32 file system `FOO.c` and `foo.C` are the same. .c should be C, .cpp should be C++.

**Comment by marc-hb** [comment]
> If we're serious about cross platform we shouldn't support patterns that don't work on some platforms. 

+1

While case-insensitivity is and endless source of issues ([security issues even](https://github.blog/2014-12-18-vulnerability-announced-update-your-git-clients/)) and should never have been implemented at the filesystem level (only at the user interface level instead), it's not going to go away and portability is one of the top reasons we're considering meson right now.

Now for the laugh: _per-directory_ case sensitivity brought to you by WSL!
https://github.com/microsoft/vscode-cmake-tools/issues/531#issuecomment-442661558
That one really made my day. Several days actually.

**Comment by scivision** [comment]
In a dream world, the filename extension as necessary metadata indicating the code language would be optional. But then there is no way to make a "magic number" as in binary files or a universally acceptable shebang.

 In the end, despite that CMake gives ways to work around this, that are like what would have been Meson `files(..., language: 'cpp')` 
I think the independent workarounds Jussi gave of using compiler flags or renaming generator are adequate and avoid breaking Meson to solve a legacy project issue. 

It's one of those hard choices that leave a few projects that work with Meson only with workarounds, to avoid making Meson messy for modern projects.

**Comment by scivision** [comment]
Recently due to bug in Intel ICL compiler (not Meson) on Windows I coincidentally had to set an equivalent flag in meson.build on a large subset of files, to override the detected language.

This wasn't too bad, and I didn't feel an absolute need for the PR declined above, though it was a cool idea.

**Comment by Volker-Weissmann** [comment]
I would also like to have a 
`map_ext_to_language(ext: '.C', language: 'cpp')`
setting, either global or project wide.
My usecase is that I have a large Codebase and I want to try meson, so I want the project to be buildable using EITHER meson OR the old build system. So I would prefer not to rename every file.

The project is Linux-only, so the fact that windows ignores the case is not a problem.

**Comment by Volker-Weissmann** [comment]
In about two releases jpakkane there will be adding a permanent solution in this repo.
Until then, you can use my fork as a temporal solution:

`git clone https://github.com/Volker-Weissmann/meson -b dot-C-files-are-C++`

It will be maintained at least until the two release cycles jpakkane talked about are over and we find a permanent solution in the upstream master branch.
("Maintaining" is easy, it is just a 4 line patch that I need to rebase every week or so)

**Comment by NotAFile** [comment]
Is this closed by https://github.com/mesonbuild/meson/pull/8747?

**Comment by mensinda** [comment]
Should be.

**Comment by Volker-Weissmann** [comment]
> Is this closed by #8747?

It is.

**Comment by Freed-Wu** [comment]
> we should definitely support files('foo.C', language : 'cpp')

> we should definitely support files('foo.C', language : 'cpp')

Any plan about it?
